oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/incrementalstore/DeltaFFSEditor.java

```diff
@@ -0,0 +1,117 @@
+    private final BufferedWriter w;
+    private final NodeStateEntryWriter entryWriter;
+    private final Predicate<String> predicate;
+    private final IncrementalFlatFileStoreStrategy incrementalFlatFileStoreStrategy;
+    private static final int LINE_SEP_LENGTH = LINE_SEPARATOR.value().length();
+
+    public DeltaFFSEditor(BufferedWriter w, NodeStateEntryWriter entryWriter, Predicate<String> predicate, IncrementalFlatFileStoreStrategy incrementalFlatFileStoreStrategy) {
+        this.w = w;
+        this.entryWriter = entryWriter;
+        this.predicate = predicate;
+        this.incrementalFlatFileStoreStrategy = incrementalFlatFileStoreStrategy;
+    }
+
+    @Override
+    public void enter(NodeState before, NodeState after) throws CommitFailedException {
+    }
+
+    @Override
+    public void leave(NodeState before, NodeState after) throws CommitFailedException {
+    }
+
+    @Override
+    public void propertyAdded(PropertyState after) throws CommitFailedException {
+    }
+
+    @Override
+    public void propertyChanged(PropertyState before, PropertyState after) throws CommitFailedException {
+    }
+
+    @Override
+    public void propertyDeleted(PropertyState before) throws CommitFailedException {
+    }
+
+    @Override
+    public @Nullable Editor childNodeAdded(String name, NodeState after) throws CommitFailedException {
+        writeToFile(after, IncrementalOperand.ADD);
+        return this;
+    }
+
+    @Override
+    public @Nullable Editor childNodeChanged(String name, NodeState before, NodeState after) throws CommitFailedException {
+        writeToFile(after, IncrementalOperand.MODIFY);
+        return this;
+    }
+
+    @Override
+    public @Nullable Editor childNodeDeleted(String name, NodeState before) throws CommitFailedException {
+        writeToFile(before, IncrementalOperand.DELETE);
+        return this;
+    }
+
+    private String getPath(NodeState e) {
+        String path;
+        if (e instanceof DocumentNodeState) {
+            path = ((DocumentNodeState)e).getPath().toString();
+        } else {
+            log.error("DeltaFFSEditor implementation is only for DocumentNodeState");
+            throw new RuntimeException("DeltaFFSEditor implementation is only for DocumentNodeState");
+        }
+        return path;
+    }
+
+    private void writeToFile(NodeState e, IncrementalOperand action) {
+        try {
+            String path = getPath(e);
+            if (!NodeStateUtils.isHiddenPath(path) && predicate.test(path)) {
+                String line =  path + "|" + entryWriter.asJson(e) + "|" + action;
+                w.append(line);
+                w.newLine();
+                incrementalFlatFileStoreStrategy.incrementEntryCount();
+                incrementalFlatFileStoreStrategy.setTextSize(incrementalFlatFileStoreStrategy.getTextSize() + line.length() + LINE_SEP_LENGTH);
+            }
+        } catch (Exception exp) {
+            log.error("Error:", exp);
```

Comment:
Error message can be a bit more detailed - 
log.error("Error while writing to the incremental FFS file :", exp);